indra/newview/llagentcamera.cpp

```diff
@@ -2145,6 +2145,13 @@ LLVector3d LLAgentCamera::calcCameraPositionTargetGlobal(BOOL *hit_limit)
     }
 // [/RLVa:KB]
 
+#ifdef OPENSIM // humbletim:FIRE-33613: Camera cannot be located at negative Z
+    // local variable shadow of the ground plane constant (used in llmax below)
+    F32 F_ALMOST_ZERO = ::F_ALMOST_ZERO;
```

Comment:
I don't really like this. Don't override a global constant with a local variable of the same name.

Reply:
Agreed in principle, though hereit is very intentional and mindfully applied to avoid upstream merge conflicts later...

Could you recommend an alternative strategy that fits in with how such changes to upstream-owned blocks of code are best maintained forward?

Reply:
From a maintainability perspective, a small reworking to declare F32 ground_plane = F_ALMOST_ZERO; the change the later line while exposing us more to the possibility of future updates, leaves a more readable code. It's not so much the override of a global (though I think we all agree that's generally ill-advised) it is more that F_ALMOST_ZERO is now (potentially) nothing like ZERO. 

I think this works, I've also added explicit "FS guards" around the opensim code change to be consistent.
If you are happy with that, please add a commit to the branch. It would do no harm to squash them into a single commit. and we can merge from there. 

```c++
// <FS:humbletim> FIRE-33613: Camera cannot be located at negative Z
F32 ground_plane = F_ALMOST_ZERO;
#ifdef OPENSIM 
    // modify the ground plancelocal variable shadow of the ground plane constant (used in llmax below)
    if (auto regionp = LLWorld::getInstance()->getRegionFromPosGlobal(camera_position_global)) 
    {
      ground_plane += regionp->getMinSimHeight();
    }
#endif
// </FS:humbletim>
    // Don't let camera go underground
    F32 camera_min_off_ground = getCameraMinOffGround();
    camera_land_height = LLWorld::getInstance()->resolveLandHeightGlobal(camera_position_global);
    F32 minZ = llmax(ground_plane, camera_land_height + camera_min_off_ground); // <FS:humbletim/> FIRE-33613: Camera cannot be located at negative Z
```

Reply:
Sounds good! Strategy implemented, squished to a single commit, and ready to go I think.

<details><summary>Additional thoughts</summary>

Leaning into readability... and as this involves a precedent-setting change (upstream assertion that `Z < 0` falls beyond shared responsibility)... it might also help to pivot around the two competing ground truths concretely (rather than as an offset):
```c++
F32 camera_ground_plane = 0.0f; // Second Life de facto ground plane 
if (OPENSIM) camera_ground_plane = OpenSimExtras.MinSimHeight; // OpenSim de facto ground plane
F32 minZ = llmax(camera_ground_plane + F_ALMOST_ZERO, ...);
```
 
But I digress. This PR fixes the immediate issue though highlights a broader challenge for OpenSim. With Second Life's limitation on negative Z-axis regions, the burden of handling these areas may shift entirely to third-party viewers like Firestorm. Since Second Life itself likely *won't be testing functionality in negative Z-axis regions*, ensuring continued functionality in OpenSim might require significant collaboration between OpenSim and TPV developers.
</details>